sql/src/main/java/io/crate/action/sql/TransportBaseSQLAction.java

```diff
@@ -247,8 +260,18 @@ public void onFailure(@Nonnull Throwable t) {
                             logger.debug("KILLED: [{}]", request.stmt());
                         } else if (Exceptions.unwrap(t) instanceof IndexShardMissingException && attempt <= MAX_SHARD_MISSING_RETRIES) {
                             logger.debug("FAILED ({}/{} attempts) - Retry: [{}]", attempt, MAX_SHARD_MISSING_RETRIES,  request.stmt());
-                            doExecute(request, listener, attempt + 1, plan.jobId());
-                            return;
+
+                            SettableFuture<Long> killFuture = SettableFuture.create();
+                            killJobs(ImmutableList.of(plan.jobId()), killFuture);
+                            try {
+                                Long numJobsKilled = killFuture.get();
```

Comment:
thread shouldn't be blocked.. use a callback. And maybe even avoid the SettableFuture by using the KillJobResponse ActionListener directly